libraries/AP_Math/control.cpp

```diff
@@ -569,45 +569,58 @@ float stopping_distance(float velocity, float p, float accel_max)
 float kinematic_limit(Vector3f direction, float max_xy, float max_z_pos, float max_z_neg)
 {
     // Reject zero-length direction vectors or undefined limits
-    if (is_zero(direction.length_squared()) || is_zero(max_xy) || is_zero(max_z_pos) || is_zero(max_z_neg)) {
+    if (is_zero(direction.length_squared())) {
+        return 0.0;
+    }
+
+    const float mag_xy = direction.xy().length();
+    
+    return kinematic_limit(mag_xy, direction.z, max_xy, max_z_pos, max_z_neg);
+}
+
+float kinematic_limit(float mag_xy, float mag_z, float max_xy, float max_z_pos, float max_z_neg)
```

Comment:
it would be good to have comments above this new version of kinematic_limit.  Also I probably wouldn't use "mag" in the argument name because that makes the reader think of magnetometers.  I think this "mag" is for "magnitude" but isn't it a length?